File: src/lib/events.ts
// Typed wrappers around Tauri listen() — one function per backend event.
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { WatcherStatus } from "./api";

export interface RecordingStarted {
  meeting_id: number;
  trigger: "auto" | "manual";
  started_at: string;
  live_captions: boolean;
}

export interface RecordingLevel {
  mic_rms: number; // 0..1
  loopback_rms: number; // 0..1
  elapsed_ms: number;
}

export interface TrackHealth {
  connected: boolean;
  device_loss_count: number;
  capture_overflow_count: number;
  capture_dropped_ms: number;
  live_dropped_ms: number;
  live_disconnected: boolean;
  fatal_error: string | null;
}

export interface RecordingHealth {
  mic: TrackHealth;
  loopback: TrackHealth;
  writer_error: string | null;
}

export type Stage = "vad" | "asr" | "diarize" | "encode";

export interface TranscriptionProgress {
  meeting_id: number;
  stage: Stage;
  pct: number; // 0..100
}

export interface TranscriptionDone {
  meeting_id: number;
  error: string | null;
}

export interface ModelDownloadProgress {
  model_id: string;
  file: string;
  file_index: number; // 1-based position within the download job
  file_count: number;
  downloaded_bytes: number;
  total_bytes: number | null;
  done: boolean;
  error: string | null;
}

const on = <T,>(event: string) => (cb: (payload: T) => void): Promise<UnlistenFn> =>
  listen<T>(event, (e) => cb(e.payload));

export const onRecordingStarted = on<RecordingStarted>("recording-started");
export const onRecordingStopped = on<{ meeting_id: number }>("recording-stopped");
export const onRecordingLevel = on<RecordingLevel>("recording-level");
export const onRecordingHealth = on<RecordingHealth>("recording-health");
export const onWatcherStatus = on<WatcherStatus>("watcher-status");
export const onTranscriptionProgress = on<TranscriptionProgress>("transcription-progress");
export const onTranscriptionComplete = on<TranscriptionDone>("transcription-complete");
export const onTranscriptionFailed = on<TranscriptionDone>("transcription-failed");
export const onModelDownloadProgress = on<ModelDownloadProgress>("model-download-progress");
export const onMeetingsChanged = on<void>("meetings-changed");

export interface LiveTranscript {
  meeting_id: number;
  track: "mic" | "loopback";
  start_ms: number;
  end_ms: number;
  text: string;
}

export const onLiveTranscript = on<LiveTranscript>("live-transcript");

export interface LiveCaptionsStatus {
  meeting_id: number;
  active: boolean;
  error: string | null;
}

export const onLiveCaptionsStatus =
  on<LiveCaptionsStatus>("live-captions-status");

/** True when the window was just shown as a tray flyout (blur dismisses). */
export const onTrayPopup = on<boolean>("tray-popup");

export interface UpdateProgress {
  downloaded: number;
  total: number | null;
}
/** Byte progress while an app update downloads (install follows on its own). */
export const onUpdateProgress = on<UpdateProgress>("update-progress");

File: src-tauri/src/events.rs
//! Event names + payload types emitted to the frontend. Keep in sync with
//! `src/lib/events.ts`.

use serde::Serialize;

pub const RECORDING_STARTED: &str = "recording-started";
pub const RECORDING_STOPPED: &str = "recording-stopped";
pub const RECORDING_LEVEL: &str = "recording-level";
pub const WATCHER_STATUS: &str = "watcher-status";
pub const TRANSCRIPTION_PROGRESS: &str = "transcription-progress";
pub const TRANSCRIPTION_COMPLETE: &str = "transcription-complete";
pub const TRANSCRIPTION_FAILED: &str = "transcription-failed";
pub const MODEL_DOWNLOAD_PROGRESS: &str = "model-download-progress";
pub const MEETINGS_CHANGED: &str = "meetings-changed";
pub const LIVE_TRANSCRIPT: &str = "live-transcript";

#[derive(Debug, Clone, Serialize)]
pub struct RecordingStarted {
    pub meeting_id: i64,
    pub trigger: &'static str,
    pub started_at: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecordingStopped {
    pub meeting_id: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecordingLevel {
    pub mic_rms: f32,
    pub loopback_rms: f32,
    pub elapsed_ms: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct WatcherStatus {
    pub enabled: bool,
    pub teams_key_found: bool,
    pub mic_in_use: bool,
    pub suppressed: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct TranscriptionProgress {
    pub meeting_id: i64,
    pub stage: &'static str, // vad | asr | diarize | encode
    pub pct: f32,
}

#[derive(Debug, Clone, Serialize)]
pub struct TranscriptionDone {
    pub meeting_id: i64,
    pub error: Option<String>,
}

/// Provisional live-caption line; replaced by the full-quality transcript
/// once the meeting ends and the pipeline runs.
#[derive(Debug, Clone, Serialize)]
pub struct LiveTranscript {
    pub meeting_id: i64,
    pub track: &'static str, // mic | loopback
    pub start_ms: u64,
    pub end_ms: u64,
    pub text: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ModelDownloadProgress {
    pub model_id: String,
    pub file: String,
    pub downloaded_bytes: u64,
    pub total_bytes: Option<u64>,
    pub done: bool,
    pub error: Option<String>,
}
